jdisc-cloud-aws/src/main/java/ai/vespa/secret/aws/AsmTenantSecretReader.java

```diff
@@ -32,31 +34,35 @@ public AsmTenantSecretReader(AsmSecretConfig secretConfig,
         super(secretConfig, identities);
         this.system = tenantConfig.system();
         this.tenant = tenantConfig.tenant();
-        this.vaultIds = createVaultIdMap(tenantConfig);
+        this.vaults = createVaultIdMap(tenantConfig);
     }
 
     // For testing
-    AsmTenantSecretReader(Function<AwsRolePath, SecretsManagerClient> clientAndCredentialsSupplier,
-                          String system, String tenant, Map<VaultName, VaultId> vaultIds) {
+    AsmTenantSecretReader(Function<AssumedRoleInfo, SecretsManagerClient> clientAndCredentialsSupplier,
+                          String system, String tenant, Map<VaultName, Vault> vaults) {
         super(clientAndCredentialsSupplier);
         this.system = system;
         this.tenant = tenant;
-        this.vaultIds = vaultIds;
+        this.vaults = vaults;
     }
 
-    static Map<VaultName, VaultId> createVaultIdMap(AsmTenantSecretConfig config) {
+    static Map<VaultName, Vault> createVaultIdMap(AsmTenantSecretConfig config) {
         // Note: we can rightfully assume that the vaults are unique by name for a tenant.
         return config.vaults().stream()
-                .map(vault -> Map.entry(VaultName.of(vault.name()), VaultId.of(vault.id())))
+                .map(vault -> Map.entry(VaultName.of(vault.name()), new Vault(VaultId.of(vault.id()), VaultName.of(vault.name()), ExternalId.of(vault.externalId()))))
                 .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
     }
 
     @Override
     protected AwsRolePath awsRole(VaultName vault) {
-        if ( ! vaultIds.containsKey(vault)) {
+        if ( ! vaults.containsKey(vault)) {
             throw new IllegalArgumentException("No vault id found for " + vault);
         }
-        return AthenzUtil.awsReaderRole(system, tenant, vaultIds.get(vault));
+        return AthenzUtil.awsReaderRole(system, tenant, vaults.get(vault).vaultId());
+    }
+
+    protected ExternalId externalId(VaultName vaultName) {
+        return Optional.ofNullable(vaults.get(vaultName)).map(Vault::externalId).orElse(null);
```

Comment:
Shouldn't this throw if externalId is empty?

Reply:
Yes, let's do in next iteration